### DISCUSSION TITLE
how to add new field in a 'zip' jagged array

### DISCUSSION THREAD

**Discussion by ZhenxuanZhang-Jensen** [scikit-hep/awkward#2017] [Q&A] [answered]
Hi experts, I want to add a new field in an already zipped jagged array. For example, if I zip 4D info into a muons object, then I can call pt,eta,phi,charge like this: `muons.Muon.pt`. However, if I want to add a new field such as 2*pt into this muons object, then I can't do this with muons.Muon['2pt'] = 2 * arrays['Muon_pt"]. Is there anything I misunderstand or how can I add a new field in this jagged array? Could you please help me? thanks
```
muons = ak.zip({
    "pt": arrays["Muon_pt"],
    "eta": arrays["Muon_eta"],
    "phi": arrays["Muon_phi"],
    "charge": arrays["Muon_charge"],
})
```

**Comment by agoose77** [comment, marked as answer]
The syntax you are describing (`muons.Muon['2pt'] = 2 * arrays['Muon_pt"]`) is not quite right for assigning to an array. It also doesn't match up with the `muons` array that you've defined above, so I will assume it's a typing mistake?

If you have an array `muons` defined as 
```python
muons = ak.zip({
    "pt": arrays["Muon_pt"],
    "eta": arrays["Muon_eta"],
    "phi": arrays["Muon_phi"],
    "charge": arrays["Muon_charge"],
})
```
then you can assign a new/existing field with 
```python
muons['pt'] = ...
```

where `...` is a Python expression.

This might fail if the expression cannot be broadcast against the existing array.

**Reply by jpivarski** [reply]
Just to keep everything cross-linked, the question was also asked on https://stackoverflow.com/q/74846508/1623645.

**Reply by ZhenxuanZhang-Jensen** [reply]
Hi Angus, actually, I did type wrong, but not the `muons.Muon`, is the 'charge' array should change to 'mass' array. Since we can separate the 'object' and 'kinematic variables' automatically with an underline when we zip 4D info with awkward(not sure why yet). for example:
```
with uproot.open(file, timeout = 1800) as f:
 tree = f["Events"]
muons = tree.arrays(['Muon_pt','Muon_eta','Muon_phi','Muon_mass'], library = "ak", how = "zip")
```
then the `muons` can call the pt,eta,phi,mass in this way automatically: `muons.Muon.pt`, `muons.Muon.eta`... For this reason, I don't know how to add a new field in this `muons.Muon`. But I found that if I assign it to a new value and reassign it to muons.Muon which can work like:
```
Muon = muons.Muon
Muon['pt2'] = ...
muons.Muon = Muon
muons.Muon.pt2
```
In this way, I can add this pt2 field in the `muons.Muon`.

**Reply by agoose77** [reply]
@ZhenxuanZhang-Jensen Oh, I learned something about `uproot`'s `zip` argument from this — it removes the common prefix of like-named ragged branches.

Back to your question.

You cannot set the value of a field with `array.field = value`. In fact, if you already have a field with the name `field`, it will raise an exception:
```pycon
>>> import uproot
>>> import skhep_testdata
>>> import awkward as ak
>>> arrays = uproot.open(skhep_testdata.data_path("uproot-HZZ.root"))["events"].arrays(how="zip")
>>> arrays.Photon = 2
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[29], line 1
----> 1 arrays.Photon = 2

File ~/Git/awkward-1.0/src/awkward/highlevel.py:1160, in Array.__setattr__(self, name, value)
   1158     super().__setattr__(name, value)
   1159 elif name in self._layout.fields:
-> 1160     raise ak._errors.wrap_error(
   1161         AttributeError(
   1162             "fields cannot be set as attributes. use #__setitem__ or #ak.with_field"
   1163         )
   1164     )
   1165 else:
   1166     raise ak._errors.wrap_error(
   1167         AttributeError(
   1168             "only private attributes (started with an underscore) can be set on arrays"
   1169         )
   1170     )

AttributeError: fields cannot be set as attributes. use #__setitem__ or #ak.with_field
```

The proper syntax for setting a field is to use the subscript operator, e.g. `array['x'] = ...`. However, for _nested_ fields, e.g. `arrays.Photon.Px`, you need to use multiple strings 
```python
arrays['Photon', 'Px'] = ...
``` 
or tuple of strings--
```python
arrays[('Photon', 'Px')] = ...
```

You can't use `arrays.Photon['Px'] = ...` — see [the user guide](https://awkward-array.org/doc/main/user-guide/how-to-restructure-add-fields.html#using-array-x) for details as to why this doesn't work.

So, if `uproot` has given you an array with a type like
```python
2421 * {
    Muon: var * {
        pt: float32,
        eta: float32,
        phi: float32,
        mass: float32
    }
}
```
So, you just need to write
```python
muons['Muon', 'pt2'] = 2 * muons.Muon.pt
```

**Reply by ZhenxuanZhang-Jensen** [reply]
Thank you so much! The answer looks very nice to me.
